### ISSUE TITLE
How to use vcglib to reduce the size of a model(<100,000 triangles)?

### ISSUE THREAD

**Issue by waseemkhan1989** [cnr-isti-vclab/vcglib#55]
Hi,

I don't know, how can I use this vcglib? My main goal is to reduce the size of a model(<100,000 triangles).
I am using Microsoft Windows 7 with Visual Studio 2017. The only way that I figured out is to compile and build the "vcglib-master\vcglib-master\apps\test\tetramesh" or "vcglib-master\vcglib-master\apps\test\trackball" in visual Studio 2017 to generate .exe file to serve my purpose. I tried this one(vcglib-master\vcglib-master\apps\test\tetramesh) but during proces of building this project, I got an error:

vcglib-master\vcglib-master\apps\test\tetramesh\test\test.cpp(1): fatal error C1083: Cannot open include file: 'simplex\vertex\with\a.h': No such file or directory

Any idea where I can find this missing file and other files? The other project is not compatible(I don't know why). As I am new to this, any help would be much appreciated.
Thanks.

Best,
Waseem Khan

**Comment by waseemkhan1989** [comment]
Hi,

I realize that there is no one answering to the questions that has been poted on this repository for more than 1 year. Should I hope to receive any response in near future?

**Comment by cignoni** [comment]
Please use the vcglib/app/tridecimator project from the devel branch. 
It compiles regularly and it does what you need (take a mesh in input and simplify it using the same algorithm inside meshlab).

**Comment by waseemkhan1989** [comment]
Hi,

Thanks for your reply. I am facing another issue. I compiled vcglib/app/tridecimator through cmake software on Windows 8.1 successfully but now I am trying to build teh project in Debug mode with win32 settings and I am getting many errors which you could see following:

1>------ Build started: Project: tridecimator, Configuration: Debug Win32 ------
1>Build started 3/16/2019 2:38:44 PM.
1>Project file contains ToolsVersion="15.0". This toolset may be unknown or missing, in which case you may be able to resolve this by installing the appropriate version of MSBuild, or the build may have been forced to a particular ToolsVersion for policy reasons. Treating the project as if it had ToolsVersion="4.0". For more information, please see http://go.microsoft.com/fwlink/?LinkId=291333.
1>Project file contains ToolsVersion="15.0". This toolset may be unknown or missing, in which case you may be able to resolve this by installing the appropriate version of MSBuild, or the build may have been forced to a particular ToolsVersion for policy reasons. Treating the project as if it had ToolsVersion="4.0". For more information, please see http://go.microsoft.com/fwlink/?LinkId=291333.
1>Project file contains ToolsVersion="15.0". This toolset may be unknown or missing, in which case you may be able to resolve this by installing the appropriate version of MSBuild, or the build may have been forced to a particular ToolsVersion for policy reasons. Treating the project as if it had ToolsVersion="4.0". For more information, please see http://go.microsoft.com/fwlink/?LinkId=291333.
1>InitializeBuildStatus:
1>  Touching "tridecimator.dir\Debug\tridecimator.unsuccessfulbuild".
1>CustomBuild:
1>  All outputs are up-to-date.
1>ClCompile:
1>  tridecimator.cpp
1>C:\Users\WASEEM\Desktop\vcglib-devel\vcglib-devel\wrap/callback.h(67): warning C4996: 'strcpy': This function or variable may be unsafe. Consider using strcpy_s instead. To disable deprecation, use _CRT_SECURE_NO_WARNINGS. See online help for details.
1>          C:\Program Files (x86)\Microsoft Visual Studio 10.0\VC\include\string.h(105) : see declaration of 'strcpy'
1>C:\Users\WASEEM\Desktop\vcglib-devel\vcglib-devel\wrap/callback.h(69): warning C4996: 'vsnprintf': This function or variable may be unsafe. Consider using vsnprintf_s instead. To disable deprecation, use _CRT_SECURE_NO_WARNINGS. See online help for details.
1>          C:\Program Files (x86)\Microsoft Visual Studio 10.0\VC\include\stdio.h(354) : see declaration of 'vsnprintf'
1>C:\Users\WASEEM\Desktop\vcglib-devel\vcglib-devel\wrap/io_trimesh/io_material.h(68): error C2864: 'vcg::tri::io::Material::index' : only static const integral data members can be initialized within a class
1>C:\Users\WASEEM\Desktop\vcglib-devel\vcglib-devel\wrap/io_trimesh/io_material.h(71): error C2864: 'vcg::tri::io::Material::Ka' : only static const integral data members can be initialized within a class
1>C:\Users\WASEEM\Desktop\vcglib-devel\vcglib-devel\wrap/io_trimesh/io_material.h(72): error C2864: 'vcg::tri::io::Material::Kd' : only static const integral data members can be initialized within a class
1>C:\Users\WASEEM\Desktop\vcglib-devel\vcglib-devel\wrap/io_trimesh/io_material.h(73): error C2864: 'vcg::tri::io::Material::Ks' : only static const integral data members can be initialized within a class
1>C:\Users\WASEEM\Desktop\vcglib-devel\vcglib-devel\wrap/io_trimesh/io_material.h(76): error C2864: 'vcg::tri::io::Material::Tr' : only static const integral data members can be initialized within a class
1>C:\Users\WASEEM\Desktop\vcglib-devel\vcglib-devel\wrap/io_trimesh/io_material.h(78): error C2864: 'vcg::tri::io::Material::illum' : only static const integral data members can be initialized within a class
1>C:\Users\WASEEM\Desktop\vcglib-devel\vcglib-devel\wrap/io_trimesh/io_material.h(79): error C2864: 'vcg::tri::io::Material::Ns' : only static const integral data members can be initialized within a class
1>C:\Users\WASEEM\Desktop\vcglib-devel\vcglib-devel\wrap/io_trimesh/import_obj.h(74): error C2864: 'vcg::tri::io::ImporterOBJ<OpenMeshType>::Info::mask' : only static const integral data members can be initialized within a class
1>          C:\Users\WASEEM\Desktop\vcglib-devel\vcglib-devel\wrap/io_trimesh/import_obj.h(68) : see reference to class template instantiation 'vcg::tri::io::ImporterOBJ<OpenMeshType>::Info' being compiled
1>          C:\Users\WASEEM\Desktop\vcglib-devel\vcglib-devel\wrap/io_trimesh/import_obj.h(1091) : see reference to class template instantiation 'vcg::tri::io::ImporterOBJ<OpenMeshType>' being compiled
1>C:\Users\WASEEM\Desktop\vcglib-devel\vcglib-devel\wrap/io_trimesh/import_obj.h(77): error C2864: 'vcg::tri::io::ImporterOBJ<OpenMeshType>::Info::cb' : only static const integral data members can be initialized within a class
1>C:\Users\WASEEM\Desktop\vcglib-devel\vcglib-devel\wrap/io_trimesh/import_obj.h(82): error C2864: 'vcg::tri::io::ImporterOBJ<OpenMeshType>::Info::numTexCoords' : only static const integral data members can be initialized within a class
1>C:\Users\WASEEM\Desktop\vcglib-devel\vcglib-devel\vcg/complex/algorithms/local_optimization/tri_edge_collapse_quadric.h(85): error C2864: 'vcg::tri::TriEdgeCollapseQuadricParameter::BoundaryQuadricWeight' : only static const integral data members can be initialized within a class
1>C:\Users\WASEEM\Desktop\vcglib-devel\vcglib-devel\vcg/complex/algorithms/local_optimization/tri_edge_collapse_quadric.h(86): error C2864: 'vcg::tri::TriEdgeCollapseQuadricParameter::FastPreserveBoundary' : only static const integral data members can be initialized within a class
1>C:\Users\WASEEM\Desktop\vcglib-devel\vcglib-devel\vcg/complex/algorithms/local_optimization/tri_edge_collapse_quadric.h(87): error C2864: 'vcg::tri::TriEdgeCollapseQuadricParameter::AreaCheck' : only static const integral data members can be initialized within a class
1>C:\Users\WASEEM\Desktop\vcglib-devel\vcglib-devel\vcg/complex/algorithms/local_optimization/tri_edge_collapse_quadric.h(88): error C2864: 'vcg::tri::TriEdgeCollapseQuadricParameter::HardQualityCheck' : only static const integral data members can be initialized within a class
1>C:\Users\WASEEM\Desktop\vcglib-devel\vcglib-devel\vcg/complex/algorithms/local_optimization/tri_edge_collapse_quadric.h(89): error C2864: 'vcg::tri::TriEdgeCollapseQuadricParameter::HardQualityThr' : only static const integral data members can be initialized within a class
1>C:\Users\WASEEM\Desktop\vcglib-devel\vcglib-devel\vcg/complex/algorithms/local_optimization/tri_edge_collapse_quadric.h(90): error C2864: 'vcg::tri::TriEdgeCollapseQuadricParameter::HardNormalCheck' : only static const integral data members can be initialized within a class
1>C:\Users\WASEEM\Desktop\vcglib-devel\vcglib-devel\vcg/complex/algorithms/local_optimization/tri_edge_collapse_quadric.h(91): error C2864: 'vcg::tri::TriEdgeCollapseQuadricParameter::NormalCheck' : only static const integral data members can be initialized within a class
1>C:\Users\WASEEM\Desktop\vcglib-devel\vcglib-devel\vcg/complex/algorithms/local_optimization/tri_edge_collapse_quadric.h(92): error C2864: 'vcg::tri::TriEdgeCollapseQuadricParameter::NormalThrRad' : only static const integral data members can be initialized within a class
1>C:\Users\WASEEM\Desktop\vcglib-devel\vcglib-devel\vcg/complex/algorithms/local_optimization/tri_edge_collapse_quadric.h(93): error C2864: 'vcg::tri::TriEdgeCollapseQuadricParameter::CosineThr' : only static const integral data members can be initialized within a class
1>C:\Users\WASEEM\Desktop\vcglib-devel\vcglib-devel\vcg/complex/algorithms/local_optimization/tri_edge_collapse_quadric.h(94): error C2864: 'vcg::tri::TriEdgeCollapseQuadricParameter::OptimalPlacement' : only static const integral data members can be initialized within a class
1>C:\Users\WASEEM\Desktop\vcglib-devel\vcglib-devel\vcg/complex/algorithms/local_optimization/tri_edge_collapse_quadric.h(95): error C2864: 'vcg::tri::TriEdgeCollapseQuadricParameter::SVDPlacement' : only static const integral data members can be initialized within a class
1>C:\Users\WASEEM\Desktop\vcglib-devel\vcglib-devel\vcg/complex/algorithms/local_optimization/tri_edge_collapse_quadric.h(96): error C2864: 'vcg::tri::TriEdgeCollapseQuadricParameter::PreserveTopology' : only static const integral data members can be initialized within a class
1>C:\Users\WASEEM\Desktop\vcglib-devel\vcglib-devel\vcg/complex/algorithms/local_optimization/tri_edge_collapse_quadric.h(97): error C2864: 'vcg::tri::TriEdgeCollapseQuadricParameter::PreserveBoundary' : only static const integral data members can be initialized within a class
1>C:\Users\WASEEM\Desktop\vcglib-devel\vcglib-devel\vcg/complex/algorithms/local_optimization/tri_edge_collapse_quadric.h(98): error C2864: 'vcg::tri::TriEdgeCollapseQuadricParameter::QuadricEpsilon' : only static const integral data members can be initialized within a class
1>C:\Users\WASEEM\Desktop\vcglib-devel\vcglib-devel\vcg/complex/algorithms/local_optimization/tri_edge_collapse_quadric.h(99): error C2864: 'vcg::tri::TriEdgeCollapseQuadricParameter::QualityCheck' : only static const integral data members can be initialized within a class
1>C:\Users\WASEEM\Desktop\vcglib-devel\vcglib-devel\vcg/complex/algorithms/local_optimization/tri_edge_collapse_quadric.h(100): error C2864: 'vcg::tri::TriEdgeCollapseQuadricParameter::QualityThr' : only static const integral data members can be initialized within a class
1>C:\Users\WASEEM\Desktop\vcglib-devel\vcglib-devel\vcg/complex/algorithms/local_optimization/tri_edge_collapse_quadric.h(101): error C2864: 'vcg::tri::TriEdgeCollapseQuadricParameter::QualityQuadric' : only static const integral data members can be initialized within a class
1>C:\Users\WASEEM\Desktop\vcglib-devel\vcglib-devel\vcg/complex/algorithms/local_optimization/tri_edge_collapse_quadric.h(102): error C2864: 'vcg::tri::TriEdgeCollapseQuadricParameter::QualityQuadricWeight' : only static const integral data members can be initialized within a class
1>C:\Users\WASEEM\Desktop\vcglib-devel\vcglib-devel\vcg/complex/algorithms/local_optimization/tri_edge_collapse_quadric.h(103): error C2864: 'vcg::tri::TriEdgeCollapseQuadricParameter::QualityWeight' : only static const integral data members can be initialized within a class
1>C:\Users\WASEEM\Desktop\vcglib-devel\vcglib-devel\vcg/complex/algorithms/local_optimization/tri_edge_collapse_quadric.h(104): error C2864: 'vcg::tri::TriEdgeCollapseQuadricParameter::QualityWeightFactor' : only static const integral data members can be initialized within a class
1>C:\Users\WASEEM\Desktop\vcglib-devel\vcglib-devel\vcg/complex/algorithms/local_optimization/tri_edge_collapse_quadric.h(105): error C2864: 'vcg::tri::TriEdgeCollapseQuadricParameter::ScaleFactor' : only static const integral data members can be initialized within a class
1>C:\Users\WASEEM\Desktop\vcglib-devel\vcglib-devel\vcg/complex/algorithms/local_optimization/tri_edge_collapse_quadric.h(106): error C2864: 'vcg::tri::TriEdgeCollapseQuadricParameter::ScaleIndependent' : only static const integral data members can be initialized within a class
1>C:\Users\WASEEM\Desktop\vcglib-devel\vcglib-devel\vcg/complex/algorithms/local_optimization/tri_edge_collapse_quadric.h(107): error C2864: 'vcg::tri::TriEdgeCollapseQuadricParameter::UseArea' : only static const integral data members can be initialized within a class
1>C:\Users\WASEEM\Desktop\vcglib-devel\vcglib-devel\vcg/complex/algorithms/local_optimization/tri_edge_collapse_quadric.h(108): error C2864: 'vcg::tri::TriEdgeCollapseQuadricParameter::UseVertexWeight' : only static const integral data members can be initialized within a class
1>C:\Users\WASEEM\Desktop\vcglib-devel\vcglib-devel\apps\tridecimator\tridecimator.cpp(186): warning C4244: 'argument' : conversion from 'double' to 'float', possible loss of data
1>
1>Build FAILED.
1>
1>Time Elapsed 00:00:01.91
========== Build: 0 succeeded, 1 failed, 0 up-to-date, 0 skipped ==========

Any idea to get rid of all these errors?

**Comment by waseemkhan1989** [comment]
@cignoni : Thank you so much for your reply and I really appreciate it. I tested tridecimator with different models but with every model, I lost all texture colors/material information after the simplification. I tried each and every option available in tridecimator but all in vain. There is also one bug i.e irrespective of which model file I give as input(e.g obj, stl etc), it always output in .ply format. I spotted the bug in tridecimator.cpp file but don't know how to fix it. It is the second last line of this file and it is "vcg::tri::io::ExporterPLY<MyMesh>::Save(mesh,argv[2]);" with the header file  #include <wrap/io_trimesh/export_ply.h>. 

The texture color/material information is really important for me. Any help in this regards would be much appreciated. Thanks

Regards,
Waseem Khan
